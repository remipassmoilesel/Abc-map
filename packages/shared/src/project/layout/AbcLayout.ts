/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { AbcView } from '../AbcView';
import { AbcTextFrame } from '../text-frame';
import { AbcScale } from '../scale';
import { AbcNorth } from '../north';

/**
 * Layouts are used to export as PDF or PNG
 */
export interface AbcLayout {
  id: string;
  name: string;
  format: LayoutFormat;
  /**
   * Layout view is relative to exported size (using format)
   */
  view: AbcView;
  /**
   * Sizes and positions of frames are relative to exported dimensions
   */
  textFrames: AbcTextFrame[];
  /**
   * Scale configuration.
   *
   * Position is relative to exported dimensions.
   */
  scale?: AbcScale;
  /**
   * North configuration.
   *
   * Position is relative to exported dimensions.
   */
  north?: AbcNorth;
}

export interface LayoutFormat {
  id: string;
  /**
   * In millimeters
   */
  width: number;
  /**
   * In millimeters
   */
  height: number;
  orientation: 'portrait' | 'landscape';
}

export class LayoutFormats {
  public static readonly A5_PORTRAIT: LayoutFormat = {
    id: 'A5_PORTRAIT',
    width: 148,
    height: 210,
    orientation: 'portrait',
  };

  public static readonly A5_LANDSCAPE: LayoutFormat = {
    id: 'A5_LANDSCAPE',
    width: 210,
    height: 148,
    orientation: 'landscape',
  };

  public static readonly A4_PORTRAIT: LayoutFormat = {
    id: 'A4_PORTRAIT',
    width: 210,
    height: 297,
    orientation: 'portrait',
  };

  public static readonly A4_LANDSCAPE: LayoutFormat = {
    id: 'A4_LANDSCAPE',
    width: 297,
    height: 210,
    orientation: 'landscape',
  };

  public static readonly A3_PORTRAIT: LayoutFormat = {
    id: 'A3_PORTRAIT',
    width: 297,
    height: 420,
    orientation: 'portrait',
  };

  public static readonly A3_LANDSCAPE: LayoutFormat = {
    id: 'A3_LANDSCAPE',
    width: 420,
    height: 297,
    orientation: 'landscape',
  };

  public static All: LayoutFormat[] = [
    LayoutFormats.A5_PORTRAIT,
    LayoutFormats.A5_LANDSCAPE,
    LayoutFormats.A4_PORTRAIT,
    LayoutFormats.A4_LANDSCAPE,
    LayoutFormats.A3_PORTRAIT,
    LayoutFormats.A3_LANDSCAPE,
  ];
}
