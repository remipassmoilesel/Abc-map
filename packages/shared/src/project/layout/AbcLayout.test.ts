/**
 * Copyright © 2023 Rémi Pace.
 * This file is part of Abc-Map.
 *
 * Abc-Map is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of
 * the License, or (at your option) any later version.
 *
 * Abc-Map is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General
 * Public License along with Abc-Map. If not, see <https://www.gnu.org/licenses/>.
 */

import { AbcLayout, LayoutFormats } from './AbcLayout';

/**
 * If this test fail, you should write a migration script then adapt it
 */
describe('AbcLayout', () => {
  it('layout formats should not change without migration', () => {
    expect(LayoutFormats.All).toMatchSnapshot();
  });

  it('layout should not change without migration', () => {
    const actual: AbcLayout = {
      id: 'test-layout-id',
      name: 'Test layout',
      format: {
        id: 'Test layout format',
        width: 800,
        height: 600,
        orientation: 'portrait',
      },
      view: {
        center: [1, 2],
        resolution: 3,
        projection: {
          name: 'EPSG:4326',
        },
        rotation: 0,
      },
      textFrames: [],
      scale: {
        x: 123,
        y: 456,
      },
    };

    expect(actual).toMatchSnapshot();
  });
});
