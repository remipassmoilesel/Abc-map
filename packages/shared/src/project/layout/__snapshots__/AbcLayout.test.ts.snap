// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AbcLayout layout formats should not change without migration 1`] = `
Array [
  Object {
    "height": 210,
    "id": "A5_PORTRAIT",
    "orientation": "portrait",
    "width": 148,
  },
  Object {
    "height": 148,
    "id": "A5_LANDSCAPE",
    "orientation": "landscape",
    "width": 210,
  },
  Object {
    "height": 297,
    "id": "A4_PORTRAIT",
    "orientation": "portrait",
    "width": 210,
  },
  Object {
    "height": 210,
    "id": "A4_LANDSCAPE",
    "orientation": "landscape",
    "width": 297,
  },
  Object {
    "height": 420,
    "id": "A3_PORTRAIT",
    "orientation": "portrait",
    "width": 297,
  },
  Object {
    "height": 297,
    "id": "A3_LANDSCAPE",
    "orientation": "landscape",
    "width": 420,
  },
  Object {
    "height": 594,
    "id": "A2_PORTRAIT",
    "orientation": "portrait",
    "width": 420,
  },
  Object {
    "height": 420,
    "id": "A2_LANDSCAPE",
    "orientation": "landscape",
    "width": 594,
  },
]
`;

exports[`AbcLayout layout should not change without migration 1`] = `
Object {
  "format": Object {
    "height": 600,
    "id": "Test layout format",
    "orientation": "portrait",
    "width": 800,
  },
  "id": "test-layout-id",
  "name": "Test layout",
  "scale": Object {
    "x": 123,
    "y": 456,
  },
  "textFrames": Array [],
  "view": Object {
    "center": Array [
      1,
      2,
    ],
    "projection": Object {
      "name": "EPSG:4326",
    },
    "resolution": 3,
    "rotation": 0,
  },
}
`;
